### ISSUE TITLE
Abstract Field Specs for Bedrock2 functions

### ISSUE THREAD

**Issue by RasmusHoldsbjergCSAU** [mit-plv/fiat-crypto#1215]
Continuing on the discussion here #1150 

I have made modifications to the Bedrock2 specs for field operations that allow for arbitrary finite fields, as opposed to the current versions that have a field modulus hard-coded, and thus allowing only for prime order fields.

The changes can be found in [this](https://github.com/RasmusHoldsbjergCSAU/fiat-crypto) fork of fiat-crypto.

The main changes are to the ```Bedrock/Specs/Field.v``` file, which has been split into two files; [```Bedrock/Specs/AbstractField.v```](https://github.com/RasmusHoldsbjergCSAU/fiat-crypto/blob/master/src/Bedrock/Specs/AbstractField.v) and [```Bedrock/Specs/PrimeField.v```](https://github.com/RasmusHoldsbjergCSAU/fiat-crypto/blob/master/src/Bedrock/Specs/PrimeField.v)

```AbstractFields.v``` is similar to the current ```Field.v```, except that it does not require the field to have prime order.
```PrimeField.v``` constructs instances of prime-order fields given a prime modulus. These instances should be similar to the ones in the current ```Field.v```.

These changes propagate outwards, but adapting the pipeline for new specs has been mostly a matter of small superficial changes.

If there is interest, I will gladly open a pull request to have this integrated into fiat-crypto.

There are a few problems, however, that perhaps should be dealt with first; namely the interaction with Rupicola is not as smooth as one could hope for.

Notice, for instance, the compiler hints [here](https://github.com/RasmusHoldsbjergCSAU/fiat-crypto/blob/master/src/Bedrock/Group/AdditionChains.v#L433) for squaring and multiplying field elements. These are highly adapted to the particular use in this particular derivation, in contrast to the [current versions](https://github.com/mit-plv/fiat-crypto/blob/master/src/Bedrock/Field/Interface/Compilation2.v#L342), which are more general. Clearly one would prefer the more general hints.

For some reason, the compiler gets stuck when I try to use the current versions of compiler hints with the new Bedrock2 specs. If anyone with insight, @andres-erbsen, @jadephilipoom or even @cpitclaudel, could comment on this, it would be much appreciated.

**Comment by andres-erbsen** [comment]
Thank you for contributing. I think everybody on our end has been unusually busy of late. I do intend to get back to this and I think we are interested.

**Comment by jadephilipoom** [comment]
Yes, that's accurate! Thanks very much for the contribution. I've taken a cursory look and am interested in merging this, and will take a closer look as soon as I have time to go into detail.

**Comment by RasmusHoldsbjergCSAU** [comment]
Thanks for the response : )

If it makes things easier, we can set up a meeting to discuss how to best go about merging this when you have time; there is no hurry.
